### ISSUE TITLE
Few features can made this software the ultimate tool.

### ISSUE THREAD

**Issue by mirajus-salehin** [safing/portmaster#151]
Portmaster is really a great piece of software. Does the all things that is advertised. It will be the ultimate tool if you consider to add the following features.
- A visual way to represent the telemetries something like monitoring graph.
- Network data usage monitoring.
- Alert if a new app tries to access the network.

**Comment by dhaavi** [comment]
Before I respond to your points individually, I'd like to point you to our [current progress page](https://safing.io/next/) and our [roadmap](https://safing.io/backlog/).

About the points you mentioned:

> * A visual way to represent the telemetries something like monitoring graph.

There is a lot more monitoring and graphs coming - we are currently working a completely fresh UI, which will include first bits of that.

> * Network data usage monitoring.

This will take a little more time to arrive, as there is some decent amount of technical background work to be done for that.

> * Alert if a new app tries to access the network.

This is a very interesting idea, and I will talk with my colleagures about that! Thank you very much for the inspiration!

**Comment by mindfulsource** [comment]
@dhaavi This project is awesome, just came here from the very nice ad from Nick, the very nice Linux YouYuber's 'The Linux Experiment' channel! I'll watch this develop with great anticipation!

> 
> > * Network data usage monitoring.
> 
> This will take a little more time to arrive, as there is some decent amount of technical background work to be done for that.
> 

Amazing if you can do this! This can help from a privacy perspective, to see just how much an app is connecting to a particular domain / IP / port, at least if you have allowed it to, and then even what time of the day or intervals it's doing it. Obviously graphs and diagrams ('Insights') giving you per-app/process stats and pattern monitoring is the holy grail. Take that, Facebook.

> > * Alert if a new app tries to access the network.
> 
> This is a very interesting idea, and I will talk with my colleagures about that! Thank you very much for the inspiration!

If this is also about blocking by default and allowing user to temporarily or permanently 'allow/deny' an app (and then per domain/ip/port), well this is the whole ['Little Snitch'](https://www.obdev.at/products/littlesnitch/index.html) [idea](https://news.ycombinator.com/item?id=22206116).

If you want to check out what AFAIK is the best GUI pop-up application firewall for Linux currently in active development, it's [this newer fork of OpenSnitch](https://github.com/gustavo-iniguez-goya/opensnitch).

**Comment by dhaavi** [comment]
We now have a first version of connection prompting in [v0.6.0](https://github.com/safing/portmaster/releases/tag/v0.6.0). The Portmaster should notify you of the update.

Graphs and insights will stake take a while. ;)

**Comment by overseeing** [comment]
I've tested it in an Ubuntu 20.04 VM and it doesn't seem to work.

- After install dependencies and the DEB (and `start` and `enable` the service), the app never finishes downloading its interface. Even after a reboot and giving it 15 minutes. It's just stuck on the animations indicating there's more to load...

- To summarise, just seems not working. Toggling between Trusted, Untrusted, Danger seemed to do default blocking of websites in Firefox...I saw a couple notifications here and there, but nothing to actually control what and how to block (with a real interactive pop-up) like OpenSnitch.

- Also, I'm not encouraged by the fact the app itself has to 'download stuff'. Do you also collect telemetry? (I don't have time to inspect the source code.) If so, that would be really, really bad - literally the opposite of why someone would use this app for their privacy. It should all be offline-level blocking functionality hard-coded into the source code. UI fanciness is nice, update checking is also nice, but this is a sensitive use case. Users should at least be able to block your Portmaster itself with some simple iptables/nftables/ufw hack, and yet Portmaster still works to block other apps/domains/IPs/ports/etc and do everything OpenSnitch can do (in terms of convenient UX).

Hope that feedback is helpful. I realise it's still early times for this app. Happy to test further.

**Comment by dhaavi** [comment]
> I've tested it in an Ubuntu 20.04 VM and it doesn't seem to work.

From the rest of the post I'm not sure if you are reporting a bug or just giving your thoughts on how the Portmaster works. If you are reporting a bug, please create a separate bug issue, as this will allow us to efficiently handle the issue.

> * After install dependencies and the DEB (and `start` and `enable` the service), the app never finishes downloading its interface. Even after a couple of reboots and about an hour of use, it seems it doesn't finish downloading all app UI elements. (Many question mark icons.)

The question marks are shown when the backend had a problem with identifying the name of an executable. I will take a look. If you want the progress tracked, please create a new issue.

> * I've tried different settings combinations in the app's deep settings and I did see a couple times a pop-up showed to block or allow a connection request, but most of the time didn't work. (Either just blocked by default or allow.) I understand it's an early feature needing a lot more testing (and feature addition such as interactive features of how long and exactly in what way to block or allow a detected connection).

Without knowing exactly which knobs you turned which way, I cannot give a useful answer to this.

> * Also, I'm not encouraged by the fact the app itself has to 'download stuff'.

The Portmaster is quite a complex system. Packing everything into one package is not possible and a waste of bandwidth. You can check the status of all resources in the version information segment on the settings page. I am happy to explain resources that are not clear what they do.

Apart from downloading resources from the update server, the Portmaster is fully self-sufficient and does _not_ require nor use anything else in order to work. Except of course for the configured DNS servers.

> Do you also collect telemetry?

No, we don't.
We do plan to collect some telemetry in the future, but it will be opt-in and you'll be able to manually review every bit before submitting.

> If so, that would be really, really bad - literally the opposite of why someone would use this app for their privacy.

We are fully aware of that. That is why we are being extra careful.

> It should all be offline-level blocking functionality hard-coded into the source code.

It is. See comment above.

> UI fanciness is nice, update checking is also nice, but this is a sensitive use case. Users should at least be able to block your Portmaster itself with some simple iptables/nftables/ufw hack, and yet Portmaster still works to block other apps/domains/IPs/ports/etc and do everything OpenSnitch can do (in terms of convenient UX).

That's the way it is.
But you don't have to block it, you can just turn off features you don't want.

> Hope that feedback is helpful. I realise it's still early times for this app. Happy to test further over time.

Great!

**Comment by dhaavi** [comment]
One more thing to add: The Portmaster is an alternative to Little/OpenSnitch, but in the sense that it is able to solve the same problem, but an alternative in the sense that it takes another approach. While it has the option for Allow/Block dialogs, we are building the Portmaster in a way that will make them obsolete for most of the use cases. Why? For most users, they are not effective - they just start clicking allow all the time because they get tired of it.

**Comment by Taza53** [comment]
![image](https://user-images.githubusercontent.com/34032665/111484194-cef34d00-8735-11eb-8dc2-2da7274eb74c.png)
if click to copy path could be added it would be helpful.

**Comment by dhaavi** [comment]
Hey @Taza53, thanks for the hints.

We already have tasks for both of these, but it might take a while until we get to them.
The issue with the spacing is also part of the bigger issue of supported screen sizes, where we are still exploring possible ways to tackle this.

In order to help us to keep track of issues here on Github, please open a new issue whenever you feel it does not fit a current issue.

**Comment by jaykhandelwal** [comment]
Really waiting for this to be added, you guys can refer to [GlassWire](https://www.glasswire.com/) if you want, it has a great implementation of this feature. Also I see that your homepage https://safing.io/portmaster/ does not contain the keyword **Firewall**, its more of a trending buzzword and might help in getting this project more traction, after all Portmaster does fall under the umbrella of Firewall

**Comment by dhaavi** [comment]
Graphs are coming in the near-ish future, although not for data usage at first. ;)

**Comment by itsnotsaved** [comment]
>A visual way to represent the telemetries something like monitoring graph.

Graph isn't important to me like data and lot of network analysis tools have graphs

> Network data usage monitoring.

It should display **per app[1]**, **per domain[2]**, **per IP [3]** basis. We must have option to view them under different timeframes otherwise difficult to read and recognize unknown/suspicious connections.

e.g. view for last 5 min, 1hr, 24hr. or specific time range e.g. from 5am to 8am
e.g. view for specific days range, from 15 Apr to 18 Apr, Jan - June

See usage field in [Glasswire](https://www.glasswire.com/) (already possible to view as it)

>Alert if a new app tries to access the network.

And portmaster need alert when existing app file modified by update etc (e.g. can display with new version number or as new software). Again, [Glasswire](https://www.glasswire.com/) already function as it. I believe this help to detect remote access tools (malicious programs and data gathering tools)

**Comment by dhaavi** [comment]
Hey @itsnotsaved, thanks for the great feedback and ideas!

We will take that into account when starting with this feature.

We are currently reworking the Network Monitor, which will lay the groundwork for so much more to come and will help us also build this feature.

**Comment by davegson** [comment]
Hey there, I just removed your last suggestion:

> Alert if a new app tries to access the network.

as this has its own suggestion now: https://github.com/safing/portmaster/issues/624

just to keep things concise

**Comment by Rexadev** [comment]
Show realtime download/upload speed of apps

**Comment by solomonrajan** [comment]
Can we expect soon?

**Comment by dhaavi** [comment]
Portmaster can now show received and sent data.

More visuals are coming soon.
